### PR TITLE
DNS support for Windows

### PR DESCRIPTION
**Pull request by msabansal**
This PR adds support for DNS for windows. In Windows the DNS resolver will listen on the gateway for the network which is routable from each container. If the name can be resolved we reply back with a positive answer. If the name can't be resolved we reply with a negative answer and the dns client in the container takes care of forwarding the queries to other DNS servers

Signed-off-by: msabansal <sabansal@microsoft.com>

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -778,6 +778,8 @@ func (c *controller) addNetwork(n *network) error {
 		return err
 	}
 
+	n.startResolver()
+
```
Comment (sanimej) [inline controller.go]: @msabansal Can a container be connected to only one network in Windows ? Even if that is the case today for nat networks I think we should consider the possibility in future a container can be connected to multiple networks. Hence the resolver instance should be associated with the sandbox and not per network. We already have platform specific files, `sandbox_dns_unix.go` and `sandbox_dns_windows.go` to abstract out the implementation details of `startResolver`.

The name/ip lookup functions are not platform specific and the service DB `svcRecords` is already keyed by the network ID. So the lookup functions `ResolveName` etc. can remain in sandbox.go and doesn't have to be refactored.

**Comment by sanimej** [comment]
`docker run` has the following dns specific config options; `--dns`, `--dns-opt` and `--dns-search`. Will any of this be used for Windows containers ? If they are not relevant will the config be blocked ?

File: endpoint.go [hunk 2]
```diff
@@ -890,6 +890,14 @@ func CreateOptionPortMapping(portBindings []types.PortBinding) EndpointOption {
...
 }
 
+// CreateOptionDNS function returns an option setter for dns entry option to
+// be passed to container Create method.
+func CreateOptionDNS(dns []string) EndpointOption {
+	return func(ep *endpoint) {
+		ep.generic[netlabel.DNSServers] = dns
+	}
+}
+
```
Comment (sanimej) [inline endpoint.go]: External nameservers are not a property of the endpoints. Since the windows DNS server will only try to resolve the container names and not doing the proxying (ie: forwarding queries to external servers) why is this necessary ?

File: resolver.go [hunk 3]
```diff
@@ -184,7 +189,14 @@ func createRespMsg(query *dns.Msg) *dns.Msg {
 }
 
 func (r *resolver) handleIPQuery(name string, query *dns.Msg, ipType int) (*dns.Msg, error) {
-	addr, ipv6Miss := r.sb.ResolveName(name, ipType)
+	var addr []net.IP
+	var ipv6Miss bool
+	if r.sb != nil {
+		addr, ipv6Miss = r.sb.ResolveName(name, ipType)
+	} else if r.network != nil {
```
Comment (sanimej) [inline resolver.go]: We should avoid this repeated if/else checks by keeping the lookup functions defined on the sandbox. Please see the earlier comment.

File: drivers/windows/windows.go [hunk 4]
```diff
@@ -421,6 +479,12 @@ func (d *driver) CreateEndpoint(nid, eid string, ifInfo driverapi.InterfaceInfo,
 		endpointStruct.IPAddress = ifInfo.Address().IP
 	}
 
+	endpointStruct.DNSServerList = strings.Join(ec.DNSServers, ",")
+
+	if n.driver.name == "nat" {
+		endpointStruct.EnableInternalDNS = true
+	}
+
```
Comment (sanimej) [inline drivers/windows/windows.go]: I believe overlay networks in Windows will not be of type nat ? What will be the behavior in that case ?
If the internal DNS can't be enabled for some reason we should check for that erro and not start the resolver. Otherwise it will result in DNS requests getting dropped.

**Comment by sanimej** [comment]
One of the use cases we have had is to be able to run a DNS server in a container and use that as the resolver for other containers. Can you confirm that it will work for Windows as well ?

**Comment by sanimej** [comment]
`In Windows the DNS resolver will listen on the gateway for the network which is routable from each container.`

If the host is already running a DNS server and using the default config it will bind port 53 on all IPs. In that case setting up the internal DNS on the gateway IP will fail. Will this be a limitation or are there any alternatives ? In Linux we explicitly setup the listener in the container namespace on a random port and do the port translation from port 53 to this random port to not break this use case.

File: controller.go (hunk 1 above)
Comment (msabansal) [inline controller.go]: The reason we have the resolver associated with the network is we can't distinguish a sandbox based on the request. There is a single resolver listening on the gateway IP of the network. The IP is routable from the container. In windows DNS servers are associated with a network adapter (endpoint). If there are multiple networks we will have multiple endpoints each with it's own list of dns servers (which will be the gateway IP plus any other user supplied dns servers).

File: endpoint.go (hunk 2 above)
Comment (msabansal) [inline endpoint.go]: We have added this to support the --dns server options. These servers will be passed on to HNS for configuring the container. In windows dns servers are specific to network adapters and are not universal as in linux

File: drivers/windows/windows.go (hunk 4 above)
Comment (msabansal) [inline drivers/windows/windows.go]: We will ensure that this flag is set. The gateway ip will be routable from container in the case of overlay network aswell.

**Comment by msabansal** [comment]
> One of the use cases we have had is to be able to run a DNS server in a container and use that as the resolver for other containers. Can you confirm that it will work for Windows as well ?
This cannot be supported with our current implementation. There is no way for us to execute a query in the context of another container (particularly for Hyperv containers which are default on developer machines)

> If the host is already running a DNS server and using the default config it will bind port 53 on all IPs. In that case setting up the internal DNS on the gateway IP will fail. Will this be a limitation or are there any alternatives ? In Linux we explicitly setup the listener in the container namespace on a random port and do the port translation from port 53 to this random port to not break this use case.

Yes that is a limitation which can't be avoided. We have no port translation facility available on windows to enable this.

File: controller.go (hunk 1 above)
Comment (sanimej) [inline controller.go]: A container connected to multiple networks should be able to resolve the service names in any network its connected to. This is done by walking through all the end-points in `resolveName` and lookup the service DB of the network to which the EP belongs.

For Windows containers if the query from the container can be associated with only one of the networks (depending on which interface the client sends the query on) is there any way to search the other networks as well ?

File: resolver.go (hunk 3 above)
Comment (sanimej) [inline resolver.go]: We can still avoid these boilerplate if/else checks in multiple places by moving the lookup functions ResolveName, ResolveIP, ResolveService to a new interface in resolver.go. These methods can be implemented by both Sandbox and Network objects. And `resolver.sb` can be changed to `resolver.backend` which can be either sandbox or the network object. This will make the resolver agnostic to how the backend lookup functions are implemented.

File: controller.go (hunk 1 above)
Comment (msabansal) [inline controller.go]: Yes. Windows DNS resolver inside the container would send out a new query for every endpoint the container is connected to so this should not be an issue. Every network connected to the container will get a query and will be able to resolve the name.

File: resolver.go (hunk 3 above)
Comment (msabansal) [inline resolver.go]: Agreed. Will make this change

Comment (msabansal) [inline resolver.go]: We do have a slight difference in implementation in network and sandbox case though. In case of networks we want to reply back with a negative response if the query fails while in case of sandbox the resolver continues executing the query from the context of the sandbox (this functionality cannot be implemented in windows at this time)

Comment (sanimej) [inline resolver.go]: We can abstract out this difference in a generic way; lets add a flag `noProxy` to the resolver structure which is a config option passed in the NewResolver call. It will be `false` in the *nix version of startResolver and `true` for Windows.

And in the ServeDNS we can check this to return `ServerFailure`.

**Comment by sanimej** [comment]
We currently support the concept of container scoped aliases; ie: one container can add an alias for another container's name using `--link option`.  In the example below only for c2, web is an alias for c1.

````
vagrant@net-1:~$ docker run -d --net docker --name c1 busybox top
83006320f67d176ff29a9d07405e2433c759a45149650039bb8930dcae915f73
vagrant@net-1:~$ docker run -d --net docker --name c2 --link c1:web busybox top
09156b7efd49196048fd5b9093107d3da4b0c58edd4a8157a2513d0f7a2cdc24
vagrant@net-1:~$ docker exec -it c2 sh
/ # nslookup c1
Server:    127.0.0.11
Address 1: 127.0.0.11

Name:      c1
Address 1: 172.18.0.2 c1.docker
/ # nslookup web
Server:    127.0.0.11
Address 1: 127.0.0.11

Name:      web
Address 1: 172.18.0.2 c1.docker
/ # 
````
I think this can't be supported in Windows because of resolver being at the network level and not per container. This should also be mentioned in the caveats with other limitations we have identified.

**Comment by msabansal** [comment]
@sanimej Incorporated the changes. 

Not sure why this test is failing: 
--- FAIL: TestNetworkRequest (0.02s)
	utils_test.go:226: exected 172.17.0.0/16. got 172.18.0.0/16

Don't think I have any changes that should impact this. I am trying to figure it out

**Comment by sanimej** [comment]
`--- FAIL: TestNetworkRequest (0.02s)`

Yes, this failure is not related to your changes. This has been fixed in the master.  If you rebase to the latest master it should pass.

File: endpoint.go (hunk 2 above)
Comment (mavenugo) [inline endpoint.go]: @msabansal its fine to have windows network adapters to have specific dns servers. But from Docker POV, the configuration is per container and this configuration is passed to libnetwork when the sandbox is created. Shouldn't we reuse that and use it per-endpoint instead of adding a new option per endpoint ?

**Comment by mavenugo** [comment]
@msabansal can you please rebase & resolve the conflicts ?

**Comment by msabansal** [comment]
@mavenugo Working on it. Getting it done by EOD

File: endpoint.go (hunk 2 above)
Comment (msabansal) [inline endpoint.go]: @mavenugo With the current libnetwork model you can have an endpoint created before the sandbox is created.  We don't allow updating endpoints after they have been created.  We need the DNS information before endpoint creation and the current model doesn't guarantee it.

File: network.go [hunk 5]
```diff
@@ -51,6 +51,18 @@ type Network interface {
...
+	// value will be true if the name exists in docker domain but doesn't have an
+	// IPv6 address. Such queries shouldn't be forwarded to external nameservers.
+	ResolveName(name string, iplen int) ([]net.IP, bool)
+	// ResolveIP returns the service name for the passed in IP. IP is in reverse dotted
+	// notation; the format used for DNS PTR records
+	ResolveIP(name string) string
+	// ResolveService returns all the backend details about the containers or hosts
+	// backing a service. Its purpose is to satisfy an SRV query
+	ResolveService(name string) ([]*net.SRV, []net.IP)
+
```
Comment (sanimej) [inline network.go]: These Resolve methods are defined in the DNSBackend interface. Please remove it from the Network interface. 

Its not necessary in the Sandbox interface either where its currently defined. That can be removed as well.

**Comment by msabansal** [comment]
@mavenugo Done. There is one test failing but I don't think its related. I will take a look at the test soon.

File: network_windows.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		hcsshim.SetCurrentThreadCompartmentId(0)
+		runtime.UnlockOSThread()
+	}()
+
+	x()
+}
+
+func (n *network) startResolver() {
+	n.resolverOnce.Do(func() {
+		log.Warnf("Launching DNS server for network", n.Name())
```
Comment (sanimej) [inline network_windows.go]: This should be log.Debugf

File: network_windows.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		hnsresponse, err := hcsshim.HNSNetworkRequest("GET", hnsid, "")
+		if err != nil {
+			log.Errorf("Resolver Setup/Start failed for container %s, %q", n.Name(), err)
+			return
+		}
+
+		for _, subnet := range hnsresponse.Subnets {
+			if subnet.GatewayAddress != "" {
+				resolver := NewResolver(subnet.GatewayAddress, false, "", n)
+				log.Warnf("Binding a resolver on network %s gateway %s", n.Name(), subnet.GatewayAddress)
```
Comment (sanimej) [inline network_windows.go]: This should be log.Debugf

File: network_windows.go (hunk 6 above)
Comment (msabansal) [inline network_windows.go]: Accepted. Will fix all the logging related stuff asap.

**Comment by sanimej** [comment]
@msabansal Other than couple of nits the changes LGTM. I will try with the changes to confirm there is no collateral for Linux. Will update by tomorrow morning.

**Comment by mavenugo** [comment]
Thanks @msabansal 

LGTM

**Comment by sanimej** [comment]
Verified on docker master with these changes. Looks ok for the linux name resolution. I think we are good to merge if you can address couple of comments on changing Warn messages to Debug.

**Comment by mavenugo** [comment]
LGTM
